turn_by_turn/xtrack_line.py

```diff
@@ -0,0 +1,170 @@
+        TypeError: If the input is not a valid ``xtrack.Line``.
+        ValueError: If no monitors are found or data is inconsistent.
+    """
+    try:
+        import xtrack as xt
+    except ImportError as e:
+        raise ImportError(
+            "The 'xtrack' package is required to convert xtrack Line objects. Install it with: pip install 'turn_by_turn[xtrack]'"
+        ) from e
+
+    if not isinstance(xline, xt.Line):
+        raise TypeError(f"Expected an xtrack Line object, got {type(xline)} instead.")
+
+    # Collect monitor names and monitor objects in order from the line
+    monitor_pairs = [
+        (name, elem)
+        for name, elem in zip(xline.element_names, xline.elements)
+        if isinstance(elem, xt.ParticlesMonitor | xt.BeamPositionMonitor)
+    ]
+    # Check that we have at least one monitor
+    if not monitor_pairs:
+        raise ValueError(
+            "No ParticlesMonitor or BeamPositionMonitor found in the Line. Please add a ParticlesMonitor or BeamPositionMonitor to the Line."
+        )
+    monitor_names, monitors = zip(*monitor_pairs)
+
+    # First check that no particles were lost during tracking. There will be trailing
+    # zeros in the data if particles were lost. This might be difficult to detect.
+    assert all(
+        mon.data.particle_id[-1] == mon.data.particle_id.max() for mon in monitors
+    ), (
+        "Some particles were lost during tracking, which is not supported by this function. "
+        "Ensure that all particles are tracked through the entire line without loss."
+    )
+
+    # Check that all monitors have the same number of turns
+    nturns_set = {mon.data.at_turn.max() + 1 for mon in monitors}
+    if len(nturns_set) != 1:
+        raise ValueError(
+            "Monitors have different number of turns, have you set the monitors with different 'start_at_turn' or 'stop_at_turn' parameters?"
+        )
+    nturns = nturns_set.pop()
+
+    # Check that all monitors have the same number of particles
+    npart_set = {len(set(mon.data.particle_id)) for mon in monitors}
+    if len(npart_set) != 1:
+        raise ValueError(
+            "Monitors have different number of particles, maybe some lost particles?"
+        )
+    npart = npart_set.pop()
+
+    # Precompute masks for each monitor and particle_id
+    monitor_particle_masks = [
+        mon.data.particle_id[:, None] == np.arange(npart)[None, :] for mon in monitors
+    ]
+
+    matrices = []
+    # Loop over each particle ID
+    for particle_id in range(npart):
+        # For each plane (e.g., 'X', 'Y'), build a DataFrame: rows=BPMs, cols=turns
+        tracking_data_dict = {}
+        for plane in TransverseData.fieldnames():
+            stacked = np.vstack(
+                [
+                    getattr(mon.data, plane.lower())[
+                        monitor_particle_masks[i][:, particle_id]
+                    ]
+                    for i, mon in enumerate(monitors)
+                ]
+            )
```

Comment:
Ideally this goes back to how it was before the previous commit, with `# fmt: off` and `# fmt: on` fencing it.